src/utils/read-rows-resumption.ts

```diff
@@ -0,0 +1,258 @@
+    this.ranges = TableUtils.getRanges(options);
+    this.rowsLimit = options.limit || 0;
+    this.hasLimit = this.rowsLimit !== 0;
+    this.rowsRead = 0;
+    this.tableStrategyInfo = tableStrategyInfo;
+    // If rowKeys and ranges are both empty, the request is a full table scan.
+    // Add an empty range to simplify the resumption logic.
+    if (this.rowKeys.length === 0 && this.ranges.length === 0) {
+      this.ranges.push({});
+    }
+  }
+
+  /**
+   * Gets the next readrows request.
+   *
+   * This function computes the next readRows request that will be sent to the
+   * server. Based on the last row key calculated by data already passed through
+   * the chunk transformer, the set of row keys and row ranges is calculated and
+   * updated. The calculated row keys and ranges are used along with other
+   * properties provided by the user like limits and filters to compute and
+   * return a request that will be used in the next read rows call.
+   *
+   * @return {protos.google.bigtable.v2.IReadRowsRequest} The request options
+   * for the next readrows request.
+   */
+  getResumeRequest(): protos.google.bigtable.v2.IReadRowsRequest {
+    const reqOpts = this
+      .tableStrategyInfo as google.bigtable.v2.IReadRowsRequest;
+
+    // Create the new reqOpts
+    reqOpts.rows = {};
+
+    // TODO: preprocess all the keys and ranges to Bytes
+    reqOpts.rows.rowKeys = this.rowKeys.map(
+      Mutation.convertToBytes
+    ) as {} as Uint8Array[];
+
+    reqOpts.rows.rowRanges = this.ranges.map(range =>
+      Filter.createRange(
+        range.start as BoundData,
+        range.end as BoundData,
+        'Key'
+      )
+    );
+
+    if (this.options.filter) {
+      reqOpts.filter = Filter.parse(this.options.filter);
+    }
+
+    if (this.hasLimit) {
+      reqOpts.rowsLimit = this.rowsLimit - this.rowsRead;
+    }
+    return reqOpts;
+  }
+
+  /**
+   * Decides if the client is going to retry a request.
+   *
+   * canResume contains the logic that will decide if the client will retry with
+   * another request when it receives an error. This logic is passed along to
+   * google-gax and used by google-gax to decide if the client should retry
+   * a request when google-gax receives an error. If canResume returns true then
+   * the client will retry with another request computed by getResumeRequest. If
+   * canResume request returns false then the error will bubble up from gax to
+   * the handwritten layer.
+   *
+   * @param {GoogleError} [error] The error that Google Gax receives.
+   * @return {boolean} True if the client will retry
+   */
+  canResume(error: GoogleError): boolean {
+    // First update the row keys and the row ranges based on the last row key.
+    const lastRowKey = this.chunkTransformer
+      ? this.chunkTransformer.lastRowKey
+      : '';
+    if (lastRowKey) {
+      spliceRanges(this.ranges, lastRowKey);
+      this.rowKeys = getRowKeys(this.rowKeys, lastRowKey);
```

Comment:
I see you're modifying the row_set here for the next request, but don't you also need to modify the `rowsLimit` for the next request? Or is that handled somewhere else?

(Aside: It seems strange to me to do this kind state modification inside `canResume`. That's not where I'd look for this, and the docstring doesn't mention anything about changing state. In Python, we use a [static helper function](https://github.com/googleapis/python-bigtable/blob/da275279a7e619e4cd3e72b10ac629d6e0e1fe47/google/cloud/bigtable/data/_async/_read_rows.py#L329) that takes in a row_set and a last seen key, and outputs a revised row_set)

Reply:
Addressing the aside, I thought it was strange that this mutation was done in canResume too, but if you don't put it here then `should not retry if all the keys are read` fails and other tests fail. This modification needs to be done before the code that previously used to live [here](https://github.com/googleapis/nodejs-bigtable/blob/b2b2fdae61865a36efc8f1f077276053f41d992f/src/table.ts#L867).

Reply:
`rowsLimit` is modified in `getResumeRequest` which is called after `canResume`. If we move the modification of `rowsLimit` to here then the `if (this.hasLimit && this.rowsLimit === this.rowsRead) {` check from before which is now in `canResume` will no longer be valid. Of course, we could just change the check to `this.rowsLimit === 0` and modify the `rowsLimit` earlier like in this position and that might be a cleaner solution, but in this case I just preserved old behaviour.

Reply:
Ok, this structure feels a bit risky to me, but it doesn't seem directly related to the changes here